File: Level of Nodes - GFG/level-of-nodes.java
//{ Driver Code Starts
import java.util.*;
import java.io.*;
import java.lang.*;

class DriverClass
{
    public static void main (String[] args) {
        Scanner sc = new Scanner(System.in);
        int t = sc.nextInt();
        
        while(t-- > 0)
        {
            ArrayList<ArrayList<Integer>> list = new ArrayList<>();
            int V = sc.nextInt();
            int E = sc.nextInt();
            for(int i = 0; i < V+1; i++)
                list.add(i, new ArrayList<Integer>());
            for(int i = 0; i < E; i++)
            {
                int u = sc.nextInt();
                int v = sc.nextInt();
                list.get(u).add(v);
                list.get(v).add(u);
            }
            int X = sc.nextInt();
            
            Solution ob = new Solution();
            
            System.out.println(ob.nodeLevel(V,list,X));
        }
    }
}
// } Driver Code Ends


/*Complete the function below*/

class Solution
{
    //Function to find the level of node X.
    int nodeLevel(int V, ArrayList<ArrayList<Integer>> adj, int X)
    {
       int[] vis = new int[V];
        Queue<Integer> q = new LinkedList<>();
        q.add(0); // Start from the first node

        int level = 0; 

        while (!q.isEmpty()) {
            int size = q.size(); // Number of nodes at the current level

            for (int i = 0; i < size; i++) {
                int curr = q.poll();

                if (curr == X) 
                    return level; 

                for (int neighbor : adj.get(curr)) {
                    if (vis[neighbor] == 0) {
                        q.add(neighbor);
                        vis[neighbor] = 1;
                    }
                }
            }

            level++; // Move to the next level
        }

        return -1; // If X is not found in the graph
    }

}

File: Level of Nodes - GFG/level-of-nodes.cpp
//{ Driver Code Starts
#include<bits/stdc++.h>
using namespace std;

// } Driver Code Ends

class Solution
{
	public:
	//Function to find the level of node X.
	int nodeLevel(int V, vector<int> adj[], int X) 
	{
	      vector<int> vis(V, 0);
        queue<int> q;
        q.push(0); // Start from the first node

        int level = 0; // Initialize the level to 0

        while (!q.empty()) {
            int size = q.size(); // Number of nodes at the current level

            for (int i = 0; i < size; i++) {
                int curr = q.front();
                q.pop();

                // Check if the current node is the target node X
                if (curr == X) {
                    return level; // Return the level when X is found
                }

                for (int neighbor : adj[curr]) {
                    if (vis[neighbor] == 0) {
                        q.push(neighbor);
                        vis[neighbor] = 1;
                    }
                }
            }

            level++; // Move to the next level
        }

        return -1; // If X is not found in the graph
	}
};

//{ Driver Code Starts.


int main()
{
    
    int t;
    cin >> t;
    while(t--)
    {
    	int V, E, X;
    	cin >> V >> E;

    	vector<int> adj[V];

    	for(int i = 0; i < E; i++)
    	{
    		int u, v;
    		cin >> u >> v;
    		adj[u].push_back(v);
    		adj[v].push_back(u);
    	}
    	cin>>X;

    	Solution obj;
    	cout << obj.nodeLevel(V, adj, X) << "\n";
    }

    return 0;
}


// } Driver Code Ends